src/engine/api.rs

```diff
@@ -17,6 +17,129 @@ pub fn next_connection_id() -> ConnectionId {
 }
 
 #[derive(Debug, PartialEq, Clone)]
+pub enum ClientMessage {
+    Consumer(ConsumerManagerMessage),
+    Producer(ProducerManagerMessage),
+    Both(ConsumerManagerMessage, ProducerManagerMessage),
+}
+unsafe impl Send for ClientMessage {}
```

Comment:
My worry wasn't that it was ugly or annoying to need the `unsafe impl`, but rather that it was necessary at all. If you have to do it, you're telling the compiler that you've made sure any synchronization issues are taken care of.

When I said it happens automatically, it derives `Send` on every compound data structure that contains no `!Send` members. See this playground example: https://is.gd/VbUEEB Raw pointers are explicitly marked as `!Send`.

My fear is that if you needed to do this, and you weren't explicitly handling any threading issues, you've introduce a thread safety bug somewhere. Something in this chain of structs/enums is marked `!Send` for some reason. I suspect it is the `SocketAddr`. It looks like it is a facade on top of the OS layer. I bet it is holding a raw pointer somewhere.